### DISCUSSION TITLE
Optuna Multinode Multi GPU training using transfromers trainer

### DISCUSSION THREAD

**Discussion by rajeshitshoulders** [optuna/optuna#5722] [Q&A]
### Expected behavior

Hi, we are trying fine-tune training using transformers_hyperparameter_search for opt-350m model  to find the best trail for Learning Rate and Batch size. we are able to run the fine-tune trainin with Ray Hyperparameter tunning, but we wanted to test it with just troch frame work, hence we converted and tested code works well with single node with 8GPU. we converted the same code to run 15xGPU using 2 nodes with 8 gpu's each using ```accelerate launch``` in the pytorch:23.12-py3

we are getting below error 
RuntimeError: Detected mismatch between collectives on ranks. Rank 3 is running collective: CollectiveFingerPrint(SequenceNumber=0, OpType=BROADC
AST, TensorShape=[3150], TensorDtypes=Long, TensorDeviceTypes=TensorOptions(dtype=float (default), device=cuda, layout=Strided (default), require
s_grad=false (default), pinned_memory=false (default), memory_format=(nullopt))), but Rank 8 is running collective: CollectiveFingerPrint(Sequenc
eNumber=0, OpType=BROADCAST, TensorShape=[3154], TensorDtypes=Long, TensorDeviceTypes=TensorOptions(dtype=float (default), device=cuda, layout=St
rided (default), requires_grad=false (default), pinned_memory=false (default), memory_format=(nullopt))).Collectives differ in the following aspe
cts:   Tensor Tensor shapes: 3150vs 3154

We tried same code using Slurm with the example https://github.com/huggingface/accelerate/blob/main/examples/slurm/submit_multinode.sh, we are getting below error

    work = default_pg.broadcast([tensor], opts)
torch.distributed.DistBackendError: NCCL error in: /opt/pytorch/pytorch/torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:1727, internal error - ple
ase report this issue to the NCCL developers, NCCL version 2.19.3
ncclInternalError: Internal check failed.
Last error:

After lot of research, we are in doubt, Optuna Hyperparameter search does work with transformers library.

we would like to know Optuna can be used with accelerate + transforms + slurm

### Environment

- Optuna version: 4.0.0
- Python version: 3.10.12
- OS: Ubuntu / Docker Container nvcr.io/nvidia/pytorch:23.12-py3
- (Optional) Other libraries and their versions: transformers==4.36.2 accelerate==0.27.2 deepspeed==0.14.0

### Error messages, stack traces, or logs

```shell
work = default_pg.broadcast([tensor], opts)
torch.distributed.DistBackendError: NCCL error in: /opt/pytorch/pytorch/torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:1727, internal error - ple
ase report this issue to the NCCL developers, NCCL version 2.19.3
ncclInternalError: Internal check failed.
Last error:
```

### Steps to reproduce

1.
2.
3.
```python
# python code
```

**Comment by rajeshitshoulders** [comment]
transformers hp code block

```python
def model_init(trial):
    gc.collect()
    torch.cuda.empty_cache()
    #torch.cuda.set_device(int(os.environ['LOCAL_RANK']))
    #dist.init_process_group(backend='nccl', init_method='env://', world_size=world_size, rank=rank)
    return prepare_model()

training_args = transformers.TrainingArguments(
        args.output_dir,
        logging_steps=1,
        save_strategy=save_strategy,
        save_steps=(steps_per_epoch * args.epochs) if args.save_model else float('inf'),
        #max_steps=steps_per_epoch * args.epochs,
        max_steps=10,
        #learning_rate=args.learning_rate,
        gradient_accumulation_steps=args.gradient_accumulation_steps,
        weight_decay=0.01,
        warmup_steps=0,
        label_names=["input_ids", "attention_mask"],
        push_to_hub=False,
        report_to="none",
        disable_tqdm=True,
        gradient_checkpointing=True,
        #per_device_train_batch_size=args.batch_size,
        eval_accumulation_steps=4,
)


trainer = transformers.Trainer(
        model=None,
        model_init=model_init,
        args=training_args,
        train_dataset=tokenized_train_dataset,
        eval_dataset=tokenized_val_dataset if args.do_eval else None,
        #compute_metrics=compute_metrics if args.do_eval else None,
        tokenizer=tokenizer,
        data_collator=data_collator,
        preprocess_logits_for_metrics=prep,
)



def hp_space(trial):
    # Define hyperparameters to be tuned
    return {
            "learning_rate": trial.suggest_float("learning_rate", 0.000001, 0.001, log=True),
            #"batch_size": trial.suggest_categorical("batch_size", [8, 16, 32, 64]),
            #"batch_size": trial.suggest_int("batch_size", 8, 128),
            "per_device_train_batch_size": trial.suggest_int("per_device_train_batch_size", 32, 256),
            #"max_steps": trial.suggest_int("max_steps", 2*10, 10*10),
        }

def run_hp_search_optuna(trainer, n_trials, direction, **kwargs):
    import optuna

    def _objective(trial):
        trainer.objective = None
        trainer.train(trial=trial)

        # memory management
        del trainer.model
        gc.collect()
        torch.cuda.empty_cache()

        # Evaluate if needed
        if getattr(trainer, "objective", None) is None:
            metrics = trainer.evaluate()
            trainer.objective = trainer.compute_objective(metrics)

        return trainer.objective


trainer.run_hp_search_optuna = run_hp_search_optuna
#best_trials = trainer.hyperparameter_search(backend='optuna',direction="maximize", storage='sqlite:///optuna-opt-study.db', hp_space=hp_space, n_trials=5)
#best_trials = trainer.hyperparameter_search(
best_trails = trainer.hyperparameter_search(
    direction="minimize",
    storage='sqlite:///optuna-opt-study.db',
    backend="optuna",
    hp_space=hp_space,
    n_trials=20,
#    compute_objective=optuna_hp_space,
)
```

Ran code:

```shell
accelerate launch --config_file ./configs/deepspeed_2node_client.yaml opt/llm_sft_optuna2.py  --model_name facebook/opt-350m --output_dir o_t --epochs ${EPOCH} --batch_size 4 --save_model False --dataset_path yizhongw/self_instruct   --do_eval True --train_size 1500 --max_length 1500
```

**Comment by nzw0301** [comment]
I'm sure that this error is mainly unrelated to optuna. The question sounds like GPU and HF libraries: accelerate or transformers. The HF repos will be a more appropriate place.

**Reply by rajeshitshoulders** [reply]
Tried different transformers version getting the same issue.

I tried the example code with https://github.com/optuna/optuna-examples/blob/main/pytorch/pytorch_distributed_simple.py with MPI getting the same error.

OMPI_MCA_btl=^openib NCCL_DEBUG=DETAIL NCCL_IB_DISABLE=1 NCCL_IB_HCA=^mlx5,^mlx4  NCCL_SOCKET_IFNAME=ens32f0 mpirun --allow-run-as-root -n 16 python optuna_pytorch_ddp.py

I used backend as nccl instead of gloo, but got this error

[rank1]: Traceback (most recent call last):
[rank1]:   File "/data/llama/llama-ft/Optuna/optuna_pytorch_ddp.py", line 176, in <module>
[rank1]:     dist.barrier()
[rank1]:   File "/root/anaconda3/envs/optuna-ddp/lib/python3.10/site-packages/torch/distributed/c10d_logger.py", line 83, in wrapper
[rank1]:     return func(*args, **kwargs)
[rank1]:   File "/root/anaconda3/envs/optuna-ddp/lib/python3.10/site-packages/torch/distributed/distributed_c10d.py", line 4159, in barrier
[rank1]:     work = group.barrier(opts=opts)
[rank1]: torch.distributed.DistBackendError: NCCL error in: ../torch/csrc/distributed/c10d/NCCLUtils.hpp:317, invalid usage (run with NCCL_DEBUG=WARN for details), NCCL version 2.21.5
[rank1]: ncclInvalidUsage: This usually reflects invalid usage of NCCL library.
[rank1]: Last error:
[rank1]: Duplicate GPU detected : rank 1 and rank 9 both on CUDA device 25000
[rank3]: Traceback (most recent call last):
[rank3]:   File "/data/llama/llama-ft/Optuna/optuna_pytorch_ddp.py", line 176, in <module>
[rank3]:     dist.barrier()
[rank3]:   File "/root/anaconda3/envs/optuna-ddp/lib/python3.10/site-packages/torch/distributed/c10d_logger.py", line 83, in wrapper
[rank3]:     return func(*args, **kwargs)
[rank3]:   File "/root/anaconda3/envs/optuna-ddp/lib/python3.10/site-packages/torch/distributed/distributed_c10d.py", line 4159, in barrier
[rank3]:     work = group.barrier(opts=opts)
[rank3]: torch.distributed.DistBackendError: NCCL error in: ../torch/csrc/distributed/c10d/NCCLUtils.hpp:317, invalid usage (run with NCCL_DEBUG=WARN for details), NCCL version 2.21.5
[rank3]: ncclInvalidUsage: This usually reflects invalid usage of NCCL library.
[rank3]: Last error:
[rank3]: Duplicate GPU detected : rank 3 and rank 11 both on CUDA device 61000
[rank0]:[W1023 08:09:33.070969364 ProcessGroupNCCL.cpp:1250] Warning: WARNING: process group has NOT been destroyed before we destruct ProcessGroupNCCL. On normal program exit, the application should call destroy_process_group to ensure that any pending NCCL operations have finished in this process. In rare cases this process can exit before this point and block the progress of another member of the process group. This constraint has always been present,  but this warning has only been added since PyTorch 2.4 (function operator())




Steps to reproduce:
conad create -n optuna-ddp python=3.10.12
pip install transformers==4.36.2 accelerate==0.27.2 deepspeed==0.14.0 evaluate==0.4.1 datasets==2.18.0 ; 
pip install git+https://github.com/aicrumb/datasettokenizer
pip install optuna dataset

Add MASTE ADDRESS in the python code 

Run with openmPI 4.1.4

OMPI_MCA_btl=^openib NCCL_DEBUG=DETAIL NCCL_IB_DISABLE=1 NCCL_IB_HCA=^mlx5,^mlx4  NCCL_SOCKET_IFNAME=ens32f0 mpirun --allow-run-as-root -n 16 python optuna_pytorch_ddp.py

Please let me know if you need any additional information

**Reply by nzw0301** [reply]
To disentangle the problems, could you remove optuna based HP search then run the code?

**Comment by corentin-ryr** [comment]
I believe I had the same issue: the different GPUs would not initialize the trainer in the same way (because the hyperparameters were not properly applied to the non-main process). In your case, the problem is that the different GPUs have different `per_device_train_batch_size`, this is probably why you get a tensor shape error.

It is now solved in Transformers. huggingface/transformers#34641
